src/rendering/vcPinRenderer.cpp

```diff
@@ -0,0 +1,230 @@
+  enum vcLoadState
+  {
+    vcLoadState_InQueue,
+    vcLoadState_Downloaded,
+    vcLoadState_Loaded,
+    vcLoadState_Failed
+  } volatile loadState; // todo: proper thread safety
+  const char *pIconURL;
+
+  vcTexture *pTexture;
+  void *pTextureData;
+  int width, height;
+};
+
+struct vcPinBin
+{
+  int count;
+  int realPins;
+  udDouble3 position; // average of group
+  vcPinIcon *pIcon;
+
+  char numbers[8];
+  vcLabelInfo label;
+};
+
+struct vcPinRenderer
+{
+  udChunkedArray<vcPinBin> pinBins;
+  udChunkedArray<vcPinIcon> pinIcons;
+
+  udMutex *pMutex;
+};
+
+void vcPinRenderer_LoadIcon(void *pUserData)
+{
+  vcPinRenderer *pPinRenderer = (vcPinRenderer*)pUserData;
+  vcPinIcon *pPinIcon = nullptr;
+
+  udLockMutex(pPinRenderer->pMutex);
+  for (size_t i = 0; i < pPinRenderer->pinIcons.length; ++i)
+  {
+    if (pPinRenderer->pinIcons[i].loadState == vcPinIcon::vcLoadState_InQueue)
+    {
+      pPinIcon = &pPinRenderer->pinIcons[i];
+      break;
+    }
+  }
+
+  udReleaseMutex(pPinRenderer->pMutex);
+
+  if (pPinIcon == nullptr) // can't happen
+    return;
+
+  pPinIcon->loadState = vcPinIcon::vcLoadState_Failed;
+
+  void *pFileData = nullptr;
+  int64_t fileLen = -1;
+  if (udFile_Load(pPinIcon->pIconURL, &pFileData, &fileLen) == udR_Success)
+  {
+    int width = 0;
+    int height = 0;
+    int channelCount = 0;
+    uint8_t *pData = nullptr;
+    pData = stbi_load_from_memory((stbi_uc*)pFileData, (int)fileLen, (int*)&width, (int*)&height, (int*)&channelCount, 4);
+    if (pData != nullptr)
+    {
+      pPinIcon->width = width;
+      pPinIcon->height = height;
+      pPinIcon->pTextureData = udMemDup(pData, sizeof(uint32_t) * width * height, 0, udAF_None);
+
+      pPinIcon->loadState = vcPinIcon::vcLoadState_Downloaded;
+
+      stbi_image_free(pData);
+    }
+
+    udFree(pFileData);
+  }
+}
+
+udResult vcPinRenderer_Create(vcPinRenderer **ppPinRenderer)
+{
+  vcPinRenderer *pPinRenderer = udAllocType(vcPinRenderer, 1, udAF_Zero);
+
+  pPinRenderer->pinBins.Init(128);
+  pPinRenderer->pinIcons.Init(32);
+  pPinRenderer->pMutex = udCreateMutex();
+
+  *ppPinRenderer = pPinRenderer;
+  return udR_Success;
+}
+
+udResult vcPinRenderer_Destroy(vcPinRenderer **ppPinRenderer)
+{
+  vcPinRenderer *pPinRenderer = (*ppPinRenderer);
+  *ppPinRenderer = nullptr;
+
+
+  for (size_t i = 0; i < pPinRenderer->pinIcons.length; ++i)
+  {
+    while (pPinRenderer->pinIcons[i].loadState == vcPinIcon::vcLoadState_InQueue)
+      udYield();
+
+    vcTexture_Destroy(&pPinRenderer->pinIcons[i].pTexture);
+    udFree(pPinRenderer->pinIcons[i].pTextureData);
+  }
+
+  udDestroyMutex(&pPinRenderer->pMutex);
+  pPinRenderer->pinBins.Deinit();
+  pPinRenderer->pinIcons.Deinit();
+  udFree(pPinRenderer);
+
+  return udR_Success;
+}
+
+bool vcPinRenderer_AddPin(vcPinRenderer *pPinRenderer, vcState *pProgramState, const char *pPinIconURL, udDouble3 position, int count)
+{
+  if (pPinIconURL == nullptr)
+    return false;
+
+  count = udMax(count, 1);
```

Comment:
This seems odd. Intended?

Reply:
Yeah, if count is `0` or negative I want it to at least count for `1`